massa-protocol-worker/src/worker_operations_impl.rs

```diff
@@ -173,30 +82,116 @@ impl ProtocolWorker {
+        }
+        self.op_batch_buffer.push_back(OperationBatchItem {
+            instant: Instant::now(),
+            node_id,
+            operations_ids,
+        });
+        Ok(())
+    }
+
+    /// On processing a batch of operation ids `op_batch` from another `node_id`
+    /// Execute the following algorithm: [redirect to GitHub](https://github.com/massalabs/massa/issues/2283#issuecomment-1040872779)
+    ///
+    ///```py
+    ///def process_op_batch(op_batch, node_id):
+    ///    ask_set = void HashSet<OperationId>
+    ///    future_set = void HashSet<OperationId>
+    ///    for op_id in op_batch:
+    ///        if not is_op_received(op_id):
+    ///            if (op_id not in asked_ops) or (node_id not in asked_ops(op_id)[1]):
+    ///                if (op_id not in asked_ops) or (asked_ops(op_id)[0] < now - op_batch_proc_period:
+    ///                    ask_set.add(op_id)
+    ///                    asked_ops(op_id)[0] = now
+    ///                    asked_ops(op_id)[1].add(node_id)
+    ///                else:
+    ///                    future_set.add(op_id)
+    ///    if op_batch_buf is not full:
+    ///        op_batch_buf.push(now+op_batch_proc_period, node_id, future_set)
+    ///    ask ask_set to node_id
+    ///```
     pub(crate) async fn update_ask_operation(
         &mut self,
         operation_batch_proc_period_timer: &mut std::pin::Pin<&mut Sleep>,
     ) -> Result<(), ProtocolError> {
         let now = Instant::now();
-        while !self.op_batch_buffer.is_empty()
-        // This unwrap is ok because we checked that it's not empty just before.
-            && now >= self.op_batch_buffer.front().unwrap().instant
-        {
-            let op_batch_item = self.op_batch_buffer.pop_front().unwrap();
-            self.on_operations_announcements_received(
-                op_batch_item.operations_ids,
-                op_batch_item.node_id,
-            )
-            .await?;
-        }
-        // reset timer
-        if let Some(item) = self.op_batch_buffer.front() {
-            operation_batch_proc_period_timer.set(sleep_until(item.instant));
-        } else {
-            let next_tick = now
-                .checked_add(self.protocol_settings.operation_batch_proc_period.into())
-                .ok_or(TimeError::TimeOverflowError)?;
-            operation_batch_proc_period_timer.set(sleep_until(next_tick));
+
+        // Reset timer to the next tick.
+        let next_tick = now
+            .checked_add(self.protocol_settings.operation_batch_proc_period.into())
+            .ok_or(TimeError::TimeOverflowError)?;
+        operation_batch_proc_period_timer.set(sleep_until(next_tick));
+
+        // Always remove the items from the buffer for processing.
+        // Note that if we cannot send an item to network,
+        // in the case that we fail to acquire a permit,
+        // the item will be left unprocessed.
+        while let Some(batch_item) = self.op_batch_buffer.pop_front() {
```

Comment:
The infinite loop was likely the result of [not removing the item](https://github.com/massalabs/massa/pull/2527/files#diff-b0e3bf7d535e3ead7e360d9fb4578e80d2918a59cba84091f854d750187095f6R134), and then only removing one item [if the send was successful](https://github.com/massalabs/massa/pull/2527/files#diff-b0e3bf7d535e3ead7e360d9fb4578e80d2918a59cba84091f854d750187095f6R182)(missed the loop essentially).

Reply:
This was not the case, the problem was not removing an item from the buffer if the calculated `ask_set` was empty. A new unit test reproduced the infinite loop.